src/Cardano/Wallet/Api/Server.hs

```diff
@@ -0,0 +1,187 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedLabels #-}
+
+module Cardano.Wallet.Api.Server
+    ( server
+    ) where
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- API handlers and server using the underlying wallet layer to provide
+-- endpoints reachable through HTTP.
+
+import Prelude
+
+import Cardano.Wallet
+    ( CreateWalletError (..)
+    , NewWallet (..)
+    , ReadWalletError (..)
+    , WalletLayer (..)
+    )
+import Cardano.Wallet.Api
+    ( Addresses, Api, Wallets )
+import Cardano.Wallet.Api.Types
+    ( ApiAddress (..)
+    , ApiT (..)
+    , ApiWallet (..)
+    , WalletBalance (..)
+    , WalletPostData (..)
+    , WalletPutData (..)
+    , WalletPutPassphraseData (..)
+    , getApiMnemonicT
+    )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( SeqState (..), defaultAddressPoolGap )
+import Cardano.Wallet.Primitive.Model
+    ( availableBalance, getState, totalBalance )
+import Cardano.Wallet.Primitive.Types
+    ( AddressState, WalletId )
+import Control.Monad.Catch
+    ( throwM )
+import Control.Monad.Trans.Except
+    ( ExceptT, withExceptT )
+import Data.Generics.Internal.VL.Lens
+    ( (^.) )
+import Data.Generics.Labels
+    ()
+import Data.Quantity
+    ( Quantity (..) )
+import Servant
+    ( (:<|>) (..), NoContent, Server, err404, err409, err501 )
+import Servant.Server
+    ( Handler (..), ServantErr (..) )
+
+server :: WalletLayer SeqState -> Server Api
+server w =
+    addresses w :<|> wallets w
+
+{-------------------------------------------------------------------------------
+                                    Wallets
+-------------------------------------------------------------------------------}
+
+wallets :: WalletLayer SeqState -> Server Wallets
+wallets w =
+    deleteWallet w
+    :<|> getWallet w
+    :<|> listWallets w
+    :<|> postWallet w
+    :<|> putWallet w
+    :<|> putWalletPassphrase w
+
+deleteWallet
+    :: WalletLayer SeqState
+    -> WalletId
+    -> Handler NoContent
+deleteWallet _ _ =
+    throwM err501
+
+getWallet
+    :: WalletLayer SeqState
+    -> WalletId
+    -> Handler ApiWallet
+getWallet w wid = do
+    (wallet, meta) <- liftHandler $ readWallet w wid
+    return ApiWallet
+        { id =
+            ApiT $ meta ^. #walletId
+        , addressPoolGap =
+            ApiT $ getState wallet ^. #externalPool . #gap
+        , balance = ApiT $ WalletBalance
+            { available =
+                Quantity $ availableBalance wallet
+            , total =
+                Quantity $ totalBalance wallet
+            }
+        , delegation =
+            ApiT $ ApiT <$> meta ^. #delegation
+        , name =
+            ApiT $ meta ^. #name
+        , passphrase =
+            ApiT $ meta ^. #passphraseInfo
+        , state =
+            ApiT $ meta ^. #status
+        }
```

Comment:
I've introduced lenses here to avoid having to put qualified imports on either the primitive or the api types since we start having name clashes a bit everywhere at this level.